QDR/6886 external status labels

qqmyers opened the pull request:
**What this PR does / why we need it**: Adds a mechanism to display an additional status label on draft datasets to aid in coordinating with an external curation workflow mechanism

**Which issue(s) this PR closes**:

Closes #

**Special notes for your reviewer**: Relates to #6886 but doesn't cover all aspects. Created for QDR - will be discussed/reviewed/potentially modified/tested there before inclusion in QDR's version. This draft PR is to make it easier for community exploration.

**Suggestions on how to test this**: Call the API to set/delete the label (suggested values such as 'Queued', 'In Process', etc. - up to 32 chars, alphanumeric and spaces allowed), verify that the label is displayed and that any existing label is removed when the version is published.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**:

qqmyers commented:
@philippconzett /all - this PR is now a working capability that allows an admin to set a defined list of states/statuses a draft dataset can be in (via an :AllowedCurationLabels setting) that can be set/removed via a UI menu (under publish data, showing only for those who can publish) or API. The only affect is to set a label that shows on the dataset/file page and dataverse/search listings and is searchable by field (e.g. entering externalstatus:* in the search box). Changes to status result in a notification and email. Any status is removed upon publication.

The intent is to allow integration with an externally defined curation process that defines states that a draft dataset can be in. (A dataset can't have more than one status (although embargoed also adds a label that is independent of this), published dataset versions can't have a status). (This PR doesn't include any push mechanism to alert an external process that a draft dataset is ready to process, but one could configure a pre-publication workflow that would trigger an external app that could then use the API to report updates as curation steps are done. For now, QDR's main use case will be curators manually changing the state via the UI.)

The images below give the basic idea (examples from QDR - label color is different and they include unrelated QDR customizations such as the use of 'Data Project' instead of 'Dataset').

![image](https://user-images.githubusercontent.com/6731983/126700233-639d56b0-214f-4172-8a29-92f2020bec45.png)
![image](https://user-images.githubusercontent.com/6731983/126700513-ace456ba-fc78-4fa6-9ed9-4928a84d7d38.png)



If this looks generally useful and consistent with the larger directions for curation support, I can add documentation, etc. to finalize this PR. Otherwise, feel free to reject/close.

philippconzett commented:
@qqmyers and QDR: Thanks for working out this PR. This will help us a lot to manage our curation workflow. I'd be happy to test this in a demo instance.
@djbrooke: Looking forward to seeing this new feature released as part of the main distribution of the Dataverse software as soon as possible.

djbrooke commented:
Thanks @qqmyers for the PR and @philippconzett for the 👍 on the need for this functionality. 

I'll spin up a branch early next week so that @TaniaSchlatter @scolapasta @jggautier @sbarbosadataverse and others on the team can get a demo and discuss next steps.

qqmyers commented:
@djbrooke - I think if you spin this up and do
curl -X PUT -d "One, Two, Three"  http://localhost:8080/api/admin/settings/:AllowedCurationLabels
it should work as advertised. 
I did a couple things to fix this branch and flyway (as of current develop branch), fix an error w.r.t. updating the label display when repeatedly setting the status, and one w.r.t. what happens w.r.t. the menu if/when you can/can't see the status choices and if the AllowedCurationLabels setting exists or not.

djbrooke commented:
Thanks @qqmyers - I'll give it a go!

djbrooke commented:
Thx @qqmyers - I got the branch spinup process working for this and I enabled the labels on the test instance. I'd like the design team to review this (I'll share the link to the EC2 instance in the Slack channel), but I have a few thoughts:

- I'm interested about the decision of adding this to the "Publish" dropdown instead of edit. I think that in an installation like Harvard's where a vast majority of datasets don't go through the curation process, having this next to the publish option would create an unnecessary decision point for many authors. In an installation like QDR's where every dataset goes through curation (right?) I can see how it makes total sense under publish. To address this for the open deposit/publish workflow we could look at implementing this at the Dataverse collection level instead of installation wide, which would allow the Harvard Dataverse Repository to specify those collections that have curation processes. This in turn would introduce the idea of setting collection-specific labels to handle the different processes of each curation team, which may be challenging to manage across the installation. Just some thoughts and worth further discussion.
- I think this would work fine with pooled curation services like the Data Curation Network -- the notifications and API could help with surfacing datasets that are candidates to enter the curation pipeline. Nice!
- Searching/browsing for curation status labels would be nice to have, especially in our recent efforts to not leave things in an unpublished state :)
- I did notice a bug in the submit for review workflow. When I try to submit a dataset for review as a contributor I see a message in the background about being unable to set the curation status, but the dataset is still submitted as expected. I'll screenshot below.

<img width="1192" alt="Screen Shot 2021-08-16 at 3 20 10 PM" src="https://user-images.githubusercontent.com/7773392/129619410-7af44426-9684-4140-8ee5-23c566c4d99d.png">

If you're willing, it would be great to discuss this on a community call as well, so that @philippconzett and others can provide feedback.

pdurbin commented:
> I'm interested about the decision of adding this to the "Publish" dropdown instead of edit.

This is an interesting point. I'd say it's easy to argue that "Return to Author" is a little odd to have under "Publish" as well. Judging from an old screenshot at https://github.com/IQSS/dataverse/issues/4139#issuecomment-336974141 we used to have "Return to Author" side by side with "Publish" like this:

<img width="497" alt="Screen Shot 2021-08-16 at 4 44 38 PM" src="https://user-images.githubusercontent.com/21006/129627440-52e35d54-b021-4c56-ac86-0173c23fae73.png">

Fast forward to modern Dataverse... since "Return to Author" is already under "Publish", I can see why @qqmyers put "Change Curation Status" below it:

<img width="292" alt="Screen Shot 2021-08-16 at 4 42 43 PM" src="https://user-images.githubusercontent.com/21006/129627505-7b8b7091-cd7e-40e9-bebe-0c51acdb8cc9.png">

The dropdown is really "Publish-related actions" since it already includes "Return to Author".

TaniaSchlatter commented:
Having labels identify the status of the dataset in the publishing workflow is fine, as it fits intended use. The "button stack" is designed for accessing the primary actions associated with publishing a dataset. As Danny describes, I can see how this kind of shortcut for setting status would be useful for curation flows. However, the functionality as demoed does not fit the intended pattern of use for the Publish button. @jggautier and I have been working on a definition of "metadata" to establish a framework for grouping related Edit actions. I don't think that curation status would quality as metadata, but will review. If curation status is not considered metadata, the "Curation status" function could go in the Edit dropdown. It would probably best be grouped with Delete Dataset – appearing just above it – as both are functions, rather than navigation, unlike the other items in the dropdown.

qqmyers commented:
Re: the error on Submit for Review - that was odd - related to having command buttons in a hidden dialog whose actions are to setExternalStatus(). They don't display, can't be clicked, and yet, just having them rendered causes the error to show immediately when the Submit for Review menu item is pressed (because setExternalStatus() gets called and the user, if they can see the submit for review button, doesn't have permission to change the status.

In any case, easy to fix - stop rendering that hidden dialog for users who can't use it.

qqmyers commented:
W.r.t. placement in the menu: I think the idea that these are really status changes in the publication process is on target. Doing something like 'Submit for Review' doesn't change this status because there is no internal process defined in Dataverse for status/states between created and published (or I guess you could say that 'In Review' is the only status/state in that process.)

The goal of having these additional labels is so that organizations and/or their apps that have an externally defined process can keep Dataverse in that loop/use Dataverse to track the status, without Dataverse having to otherwise participate in that process, e.g. by moving things between states based on actions in Dataverse. I think that's why QDR thought the publish menu makes sense - this isn't metadata, isn't a change that will in any way be reflected in the published dataset (which the edit menu items do affect.) (Since QDR has a fork, the decision for the PR doesn't have to be the same, but it was the end users (curation team) rather than me as the developer who thought publish was the better place.)

One other thing to note in considering where this goes/the overall functionality, right now the ability to set the curation statuses is limited to those who can publish, as is the ability to see the curation labels. I.e. someone who has to hit Submit For Review won't ever see the curation status label in the display. That doesn't affect menu placement directly but the fact that this isn't a publicly visible status might affect how we think about these statuses (and may be something that other institutions will want to change, i.e. so that the submitters can see what state their submission is in.

W.r.t. whether this should be per Dataverse collection: the original design was for API only (to let external processes change the status). When it was API only, we started with letting the user (with ~admin privs) set an arbitrary label, but that opened the door for misuse and was harder to manage in the UI than going to a pick list. If we want per-collection options to have curation statuses, the next question is whether they all use the same pick list (one global setting) or if different subcollections need to have their own lists (with what security/permission measures?) or if collection admins should select which statuses are active in their collection from a master list set by the setting, etc. Or have a setting that lets you define multiple lists and collection admins pick one list - corresponding to their process, etc. If the choice is just to have a boolean on/off, I think I could add it to this PR. If the need is for something more complex, I'd suggest we break that into a new issue/PR.

qqmyers commented:
FWIW: Just added an api call to get the list of unpublished datasets and their curation status. This is a modification of a similar call QDR has: at QDR, we've created a custom role 'assignee' that is intended to be assigned to one curator per dataset, and the report in the api call also includes a column with that 'assignee's name.  While that might also be of community interest, it depends on a custom role and correct use of it, so for this PR I just dropped the column and left the capability to get a list of unpublished datasets, their creation dates, their last modified date, and the curation status in a csv file.

djbrooke commented:
Thanks @qqmyers for discussing earlier.

@adam3smith: a quick question for you, mostly related to the last paragraph of Jim's comment above (https://github.com/IQSS/dataverse/pull/7967#issuecomment-901453311). In order to support use cases at the Harvard Dataverse Repository and in other cases where different Dataverse collections in an installation have different curation policies (and therefore labels), I'd be interested in adding the ability for a superuser to define this on a per-collection level. Since this is potentially outside the QDR use case, I wanted to check with you -- is it OK if @qqmyers adds this functionality as part of this PR so that it can be accepted into the core code? If it's not possible to add this we can investigate other options to add this functionality outside of this PR, but it'd be helpful to get it added all at once. Let me know what you think and thanks for the PR!

adam3smith commented:
Thanks for checking. Yes, that makes perfect sense -- we're happy to make this more widely usable and of course this can be added to this PR.

djbrooke commented:
Thanks @adam3smith !!

djbrooke commented:
@qqmyers - I'll drop off this issue for now but please let me know if you need anything. Thanks!

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:
```diff
@@ -180,30 +185,31 @@
+    @Inject
+    DataverseSession session;
 
     @EJB
     DatasetServiceBean datasetService;
 
     @EJB
     DataverseServiceBean dataverseService;
-    
+
```
This file appears to have been completely reformatted, which is suboptimal. Usually we try to reformat just the code we touch: https://guides.dataverse.org/en/5.6/developers/coding-style.html#format-code-you-changed-with-netbeans

I still dream of us adopting a tool to format our code for us like gofmt or rustfmt: https://groups.google.com/g/dataverse-dev/c/y2Jpk3szTf8/m/rckKmP6-BgAJ

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/SettingsWrapper.java:
```diff
@@ -373,5 +379,20 @@ public String getDefaultMetadataLanguage() {
+
+    public List<String> getAllowedExternalStatuses(Dataset d) {
+        String setName = d.getEffectiveCurationLabelSetName();
+        logger.info("Set name: " + setName);
```
Lower to logger.fine or remove.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:
```diff
@@ -1752,217 +1770,243 @@ public Response returnToAuthor(@PathParam("id") String idSupplied, String jsonBo
+    @PUT
+    @Path("{id}/setExternalCurationStatus")
+    public Response setExternalCurationStatus(@PathParam("id") String idSupplied, @QueryParam("label") String label) {
+        logger.info("Label is " + label);
```
Reduce to logger.fine or remove.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:
```diff
@@ -2549,19 +2589,103 @@ public Response resetFileStore(@PathParam("identifier") String dvIdtf,
+            return error(Response.Status.NOT_FOUND, "No such dataset");
+        }
+
+        return response(req -> ok(dataset.getEffectiveCurationLabelSetName()));
```
I'm not sure how I feel about the lack of authentication on this GET method. We want it to work on drafts, for example?

qqmyers replied:
FWIW: This is nominally just the name of the set that's allowed, not the particular state/curation label on this dataset, but I added auth - not sure there's a reason anyone but superuser needs to know this.

pdurbin replied:
Perhaps a curator would like to know but I think superuser is better than no auth at all. Thanks.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:
```diff
@@ -2549,19 +2589,103 @@ public Response resetFileStore(@PathParam("identifier") String dvIdtf,
+        } catch (WrappedResponse ex) {
+            return error(Response.Status.BAD_REQUEST, "Authentication is required.");
+        }
+        if (!user.isSuperuser()) {
+            return error(Response.Status.FORBIDDEN, "Superusers only.");
+        }
```
I'm a little confused. The sets are defined on dataverse collections but only superusers can define the sets? Shouldn't admins of the collections be able to define the sets and otherwise manage them (including deleting them)?

djbrooke replied:
@qqmyers and I discussed, and I said I was OK with this for now. We can revisit opening this up later if it's widely adopted and/or becomes a support burden.

pdurbin replied:
Fair enough but the docs and release notes could certainly use some clarification on who can do what when. It wasn't until I got the GUI working (thanks, Jim), that it's obvious that curators/admins in the GUI cannot set up the sets. They can only choose from predefined sets:

![Screen Shot 2021-09-27 at 3 31 56 PM](https://user-images.githubusercontent.com/21006/134973837-032e11e8-7d18-43f8-a72b-ba9c4f98cf7a.png)

The curators/admins in the GUI also cannot preview what's in the sets. The sysadmin knows that the following was used to create the sets in this case:

```
{
  "Standard Process": [
    "Author contacted",
    "Privacy Review",
    "Awaiting paper publication",
    "Final Approval"
  ],
  "Alternate Process": [
    "State 1",
    "State 2",
    "State 3"
  ]
}
```

qqmyers replied:
Better labels would probably help - e.g. including an org. name that matches the info in their collection and their current curation process. If we also have the use case where people will be selecting an unknown process to start using, then showing the labels, having more documentation about their meaning, flow diagram for allowed transitions, etc. would all be important.

TaniaSchlatter replied:
Does the model for implementing this require the use of a header? What if it is desirable to have a list of straight-forward, not grouped options, as was shown in the demo?

qqmyers replied:
There may be some confusion: The UI in https://github.com/IQSS/dataverse/pull/7967#issuecomment-885191872 doesn't change - only one set of labels shows up there (no header). Only the new UI for selecting which set will show up for a given dataset is where we either show just the set labels (as is, shown above) or we also provide more info about the items available in a set (as suggested). In this new mgmt UI, the only choices are which set to choose - the individual items are not selectable/ are only informational.

TaniaSchlatter replied:
What if curators need to see the list of what is under the headers to make the correct choice?

djbrooke replied:
Hi - I'd be happy with this being managed through the API only, for superusers only, settable at the collection level or installation level. 

If we do want to implement this in the UI at the collection level, we should make it similar to the S3 selection where it can only be seen by superusers. I think the management of this process should be a coordination/negotiation between the superuser(s) of the installation and the admin(s) of a collection, and that it's important for a superuser to control which collections have what labels enabled, especially once we get into interoperation with other systems.

qqmyers replied:
I think this is resolved. It was unclear that the Dataverse General Info/Edit API only shows the choices to superusers.

pdurbin replied:
Right. When I started this thread I confused, thinking that curators/admins would be selecting the curation label set. From what I understand now, this is done by superusers who have been working closely with sysadmins that have created the sets in the first place. Of course, new superusers will be onboarded over time so I think longer term it would be nice to see the details (the curation steps or states) in the GUI.

djbrooke replied:
Thanks all for the discussion here. If this is widely adopted we can revisit by adding more details and/or making it available to other roles than superusers. I'll resolve this conversation.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:
```diff
@@ -2698,10 +2819,78 @@ public Response addFilesToDataset(@PathParam("id") String idSupplied,
+                    assignees.get(ra.getRole().getAlias()).add(ra.getAssigneeIdentifier());
+                }
+            }
+            String name = dataset.getCurrentName().replace("\"", "\"\"");
```
Since we're creating a CSV will datasets titles with commas be a problem?

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/SetExternalStatusCommand.java:
```diff
@@ -0,0 +1,118 @@
+        } else {
+            String setName = getDataset().getEffectiveCurationLabelSetName();
+            if(setName.equals(SystemConfig.CURATIONLABELSDISABLED)) {
+                logger.info("External status labeling disbaled for dataset id: " + getDataset().getId());
```
Typo and this should be logger.fine.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/SetExternalStatusCommand.java:
```diff
@@ -0,0 +1,118 @@
+                }
+            }
+            if(!found) {
+                logger.info("Label not found: " + label + " in set " + setName);
```
logger.fine please

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java:
```diff
@@ -466,7 +466,18 @@ Whether Harvesting (OAI) service is enabled
-        SendNotificationOnDatasetCreation
+        SendNotificationOnDatasetCreation,
+        /**
+         * A comma separated list of allowed labels (up to 32 characters, spaces
```
Wait, it is a comma separated list or a JSON array?

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/util/SystemConfig.java:
```diff
@@ -117,6 +123,9 @@
     
     private static final long DEFAULT_GUESTBOOK_RESPONSES_DISPLAY_LIMIT = 5000L; 
     
+    public final static String DEFAULTCURATIONLABELSET = "DEFAULT";
+    public final static String CURATIONLABELSDISABLED = "DISABLED";
```
Do these English words appear in the UI anywhere? If so, let's make sure they can be translated.

qqmyers replied:
default isn't shown/there's already a bundle entry from when storage drivers were made configurable, but disabled hadn't been handled yet. Now done.

pdurbin commented on src/main/webapp/search-include-fragment.xhtml:
```diff
@@ -545,6 +545,7 @@
                             <h:outputText value="#{bundle['dataset.versionUI.inReview']}" styleClass="label label-success" rendered="#{result.inReviewState}"/>
                             <h:outputText value="#{bundle['dataset.versionUI.unpublished']}" styleClass="label label-warning" rendered="#{result.unpublishedState}"/>
                             <h:outputText value="#{bundle['dataset.versionUI.deaccessioned']}" styleClass="label label-danger" rendered="#{result.deaccessionedState}"/>
+                            <h:outputText value="#{result.externalStatus}" styleClass="label label-info" rendered="#{!empty result.externalStatus and SearchIncludeFragment.canPublishDataset(result.entityId)}"/>
```
I'm just putting this comment at the end (where I look for them)... some tests would be nice.

qqmyers replied:
I added some tests of setting the curation label set on a dataverse and then trying to set a label from that set on a dataset in the dataverse. Hopefully the tests will pass - awaiting the automated results.

pdurbin replied:
Fantastic! Thanks for adding tests!

djbrooke replied:
+1 for tests

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/search/SearchFields.java:
```diff
@@ -220,6 +220,8 @@
      */
     public static final String PUBLICATION_STATUS = "publicationStatus";
+    
+    public static final String EXTERNAL_STATUS = "externalStatus";
```
Can you facet on this new externalStatus Solr field?

qqmyers replied:
You can't. There is the superuser API call to get the CSV table of all draft datasets and their current labels. I hope that faceting to select draft-only within a given sub-collection will make things usable without having to also facet here but it could be added (presumably listing the labels from all sets if you're looking at collections that use different sets?)

pdurbin reviewed (changes requested):
It looks like bc066f7 fixed a couple of the things I found but I left a variety of feedback.

For @TaniaSchlatter and others it would be nice to have screenshots of the new GUI additions.

TaniaSchlatter commented:
It would be a good practice to indent the states and list them under the heading. Who knows when someone is going to use what is implemented as a model for another feature. 

Is there a reason not to list the states in the drop down? Seems like low-hanging happy UI fruit.

qqmyers commented on conf/solr/8.8.1/schema.xml:
```diff
@@ -162,6 +162,7 @@
     <field name="dvSubject" type="string" stored="true" indexed="true" multiValued="true"/>
 
     <field name="publicationStatus" type="string" stored="true" indexed="true" multiValued="true"/>
+    <field name="externalStatus" type="string" stored="true" indexed="true" multiValued="false"/>
```
done

djbrooke replied:
Added some content inspired by the 5.4 notes in https://github.com/IQSS/dataverse/pull/7967/commits/8a8f51e5ff69d9d2be1f96e0f4b5e162bb7f8a22

djbrooke replied:
I'm going to resolve this issue, we can revisit at release time if needed.

pdurbin replied:
I don't think 8a8f51e is quite what's needed but yes, we can sort it out later.

I found this by luck in the 4.2 release notes:

update the solr schema.xml file
stop solr
remove current schema.xml
replace with the 4.2 schema.xml file provided
start solr

Something more like that is what's needed. That is, we typically upload a schema.xml file as an artifact to the release.

djbrooke replied:
Thx @pdurbin - I see something similar in https://github.com/IQSS/dataverse/releases/tag/v4.15 ? 

`Replace Solr schema.xml to allow sorting and filtering on the file page
-stop solr instance (service solr stop, depending on solr installation/OS, see http://guides.dataverse.org/en/4.15/installation/prerequisites.html#solr-init-script)
-replace schema.xml

cp /tmp/dvinstall/schema.xml /usr/local/solr/solr-7.3.1/server/solr/collection1/conf
cp /tmp/dvinstall/solrconfig.xml /usr/local/solr/solr-7.3.1/server/solr/collection1/conf
-start solr instance (service solr start, depending on solr/OS)`

If this looks more correct I can adapt this and get it into this branch...

pdurbin replied:
Yes, the 4.15 example is a good one. Nice find.

pdurbin replied:
Oh but in this case people only need to update schema.xml, not solrconfig.xml.

djbrooke replied:
Cool, I'll get something together. Thank you!

djbrooke replied:
Added in https://github.com/IQSS/dataverse/pull/7967/commits/fa4f8cc735417ad3948736caa7ee015a22bbd942 - will revisit at release time if needed.

qqmyers commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2385,4 +2385,16 @@ Scripts that implement this association for specific service protocols are maint
 
+.. _:AllowedCurationLabels:
+
+:AllowedCurationLabels
```
done

qqmyers commented on src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java:
```diff
@@ -198,6 +198,11 @@ public Dataset find(Object pk) {
     }
 
+    //Used in datasets listcurationstatus API
+    public List<Dataset> findAllUnpublished() {
```
I was as well, but didn't find it.

pdurbin replied:
🤷

qqmyers commented on src/main/java/edu/harvard/iq/dataverse/DataversePage.java:
```diff
@@ -1199,15 +1205,56 @@ public String getDefaultStorageDriverLabel() {
     
-    public Set<Entry<String, String>> getMetadataLanguages() {
-        return settingsWrapper.getMetadataLanguages(this.dataverse).entrySet();
+    public Set<Entry<String, String>> getCurationLabelSetOptions() {
```
Yep - a merge issue, now fixed.

qqmyers commented on src/main/webapp/dataverse.xhtml:
```diff
@@ -160,6 +160,19 @@
+                                            </h:selectOneMenu>
+                                            <p:message for="dataverseCurationLabels" display="text"/>
+                                </div>
+                            </div>
```
Yep - merge issue. Fixed.

qqmyers commented:
I think this PR is back up-to-date w.r.t. feedback. If I missed something, let me know.

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2385,4 +2385,17 @@ Scripts that implement this association for specific service protocols are maint
+ 
+A JSON Object containing lists of allowed labels (up to 32 characters, spaces allowed) that can be set, via API or UI by users with the permission to publish a dataset. The set of labels allowed 
+for datasets can be selected by a superuser - via the Dataverse collection page (Edit/General Info) or set via API call. 
+The labels in a set should correspond to the states in an organization's curation process and are intended to help users/curators track the progress of a dataset through an externally defined curation process. 
```
We probably don't need the word "externally" here. I get that integration is possible, but it's not required.

qqmyers replied:
I removed it. FWIW: The word is supposed to indicate that there really is no curation process going on in Dataverse - this is only a label and if you have to define it elsewhere, and enforce it's restrictions (such as what states can transfer to what state) elsewhere (manual or other app).

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2385,4 +2385,17 @@ Scripts that implement this association for specific service protocols are maint
+This functionality is disabled when this setting is empty/not set.
+Each set of labels is identified by a curationLabelSet name and a JSON Array of the labels allowed in that set.
+
+``curl -X PUT -d '{"Standard Process":["Author contacted", "Privacy Review", "Awaiting paper publication", "Final Approval"], "Alternate Process":["State 1","State 2","State 3"]}' http://localhost:8080/api/admin/settings/:AllowedCurationLabels``
```
All these API docs are great but I'm feeling like it would be nice if this feature were described at a higher level in the User Guide (the "Dataset + File Management" section, I guess). We could take a lot of the content from the release note (which is pretty ephemeral, which the guides are lasting). Then, of course, we would link to the details in the API Guide. Since this is a high-level docs thing, let me ping @djbrooke for thoughts.

djbrooke replied:
Sounds good, @qqmyers OK if I take a pass at this?

qqmyers replied:
Sure

djbrooke replied:
OK, added some additional content in https://github.com/IQSS/dataverse/pull/7967/commits/ae61cbeb53b1239978f487be34ecde69b24def23

pdurbin reviewed (changes requested):
Overall this is looking much better but I left some comments for things to fix before it's ready for QA.

qqmyers commented on src/test/java/edu/harvard/iq/dataverse/api/DatasetsIT.java:
```diff
@@ -2368,6 +2371,59 @@ public void testReCreateDataset() {
+        //Set curation label set on dataverse
+        //Valid option, bad user
+        Response setDataverseCurationLabelSetResponse = UtilIT.setDataverseCurationLabelSet(dataverseAlias, apiToken, "AlternateProcess");
+        setDataverseCurationLabelSetResponse.then().assertThat().statusCode(FORBIDDEN.getStatusCode());
```
This should be fixed, but I don't know that I can get to the test results so let me know if not. (I can spin up a test instance to run the IT tests, but I hope these are simple enough to not have to iterate on.)

qqmyers commented on doc/sphinx-guides/source/admin/dataverses-datasets.rst:
```diff
@@ -61,6 +61,33 @@ The available drivers can be listed with::
+Datasets within a given Dataverse collection can be annotated with a Curation Label from one of the label sets defined by :ref:`:AllowedCurationLabels <:AllowedCurationLabels>`. The allowed set can be specified via the API as shown below, or by editing the 'General Information' for a Dataverse collection on the Dataverse collection page. Only accessible to superusers. ::
+
+The curationLabelSet to use within a given collection can be set by specifying its name using::
+ 
```
+1 for getting #8117 fixed!

qqmyers commented:
Should be all set - the only remaining test fail was in an unrelated one reading http://schema.datacite.org/meta/kernel-4.1/metadata.xsd which must have been down.

djbrooke commented:
Thanks @qqmyers for the updates and thanks @pdurbin for the detailed review! I've resolved all the outstanding comments and I will move this to QA.

djbrooke commented:
Thanks @kcondon and @qqmyers for working through the last few issues! I'll integrate the release note content into #8128 !
